server/console_group.go

```diff
@@ -0,0 +1,475 @@
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, "Requires a valid group ID.")
+	}
+
+	err = UpdateGroup(ctx, s.logger, s.db, groupID, uuid.Nil, uuid.Nil, in.Name, in.LangTag, in.Description, in.AvatarUrl, in.Metadata, in.Open, int(in.MaxCount.Value))
+	if err != nil {
+		return nil, err
+	}
+
+	return &emptypb.Empty{}, nil
+}
+
+func (s *ConsoleServer) GetMembers(ctx context.Context, in *console.GroupId) (*api.GroupUserList, error) {
+	groupID, err := uuid.FromString(in.Id)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, "Requires a valid group ID.")
+	}
+
+	users, err := ListGroupUsers(ctx, s.logger, s.db, s.tracker, groupID, 0, nil, "")
+	if err != nil {
+		return nil, status.Error(codes.Internal, "An error occurred while trying to list group members.")
+	}
+
+	return users, nil
+}
+
+func (s *ConsoleServer) DemoteGroupMember(ctx context.Context, in *console.ChangeGroupUserStateRequest) (*emptypb.Empty, error) {
+	userID, err := uuid.FromString(in.Id)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, "Requires a valid user ID.")
+	}
+	groupID, err := uuid.FromString(in.GroupId)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, "Requires a valid group ID.")
+	}
+
+	if err = demoteGroupUser(ctx, s.logger, s.db, s.router, uuid.Nil, groupID, userID); err != nil {
+		if err == ErrEmptyMemberDemote {
+			return nil, status.Error(codes.FailedPrecondition, "Cannot demote user in the group.")
+		}
+		return nil, status.Error(codes.Internal, "An error occurred while trying to demote the user in the group.")
+	}
+	return &emptypb.Empty{}, nil
+}
+
+func (s *ConsoleServer) PromoteGroupMember(ctx context.Context, in *console.ChangeGroupUserStateRequest) (*emptypb.Empty, error) {
+	userID, err := uuid.FromString(in.Id)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, "Requires a valid user ID.")
+	}
+	groupID, err := uuid.FromString(in.GroupId)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, "Requires a valid group ID.")
+	}
+
+	if err = promoteGroupUser(ctx, s.logger, s.db, s.router, uuid.Nil, groupID, userID); err != nil {
+		if err == ErrEmptyMemberPromote {
+			return nil, status.Error(codes.FailedPrecondition, "Cannot promote user in the group.")
+		}
+		return nil, status.Error(codes.Internal, "An error occurred while trying to promote the user in the group.")
+	}
+
+	return &emptypb.Empty{}, nil
+}
+
+func demoteGroupUser(ctx context.Context, logger *zap.Logger, db *sql.DB, router MessageRouter, caller uuid.UUID, groupID uuid.UUID, uid uuid.UUID) error {
+	myState, tx, err := preGroupUserStateChangeValidate(ctx, logger, db, groupID, caller)
+	if err != nil {
+		return err
+	}
+	channelID, stream, err := prepareGroupStream(logger, groupID)
+	if err != nil {
+		return err
+	}
+
+	var message *api.ChannelMessage
+	ts := time.Now().Unix()
+	if err := ExecuteInTx(ctx, tx, func() error {
+		if uid == caller {
```

Comment:
Check this before opening a transaction if possible, this seems like an input validation element that can be done before any real work.